Show mini urlbar in `BraveMultiContentsView`

simonhong opened the pull request:
Show mini urlbar in BraveMultiContentsView

fix https://github.com/brave/brave-browser/issues/46161

TEST=`SplitViewLocationBarBrowserTest.*`

1. launch browser with --enable-features=SideBySide
2. Select Add tab to split view from tab's context menu
3. Check mini urlbar is displayed on the inactive tab's contents

![image](https://github.com/user-attachments/assets/2379b3e8-a20d-4bb0-9c6e-391100cfa2b9)

<!-- Add brave-browser issue below that this PR will resolve -->
Resolves

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

<!--
## Checklist:

- Review design docs
  [Browser design principles](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/chrome_browser_design_principles.md)
  [Style guide](https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/c++.md)
  [Core principles](https://www.chromium.org/developers/core-principles/)
- Ensure there are (tests)[https://www.chromium.org/developers/testing/]. Unit test as much as possible (including edge cases), but also include browser tests covering high level functionality.
- Ensure that there are comments explaining what classes/methods are/do. The "why" is often more important than the "what" in comments. Also update any relevant docs (moving docs from wiki to brave-core if necessary).
- Request security or other review (third-party libraries, rust code, etc...) if applicable [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) [other review](https://github.com/brave/reviews/issues/new/choose)
  Also see [adding third-party libraries](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/adding_to_third_party.md) for general guidelines on using third party code
- Maks sure there is a [ticket](https://github.com/brave/brave-browser/issues) for your issue
- Use Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- Write a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- Squash any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- Add appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- Run `git rebase master` (if needed)
-->

zeusshomaly commented:
@simonhong Exactly 100% what I requested! This is amazing! Tysm! After this, I hope work on the new sidebar is underway with some screenshots :) cheers Simon, you are an amazing man!

fallaciousreasoning commented on browser/ui/views/frame/split_view/brave_multi_contents_view.cc:
```diff
@@ -79,12 +90,39 @@ void BraveMultiContentsView::Layout(PassKey) {
+  int inactive_index = active_index_ == 0 ? 1 : 0;
+  if (!secondary_location_bar_) {
+    secondary_location_bar_ = std::make_unique<SplitViewLocationBar>(
+        browser_view_->browser()->profile()->GetPrefs(),
```
sad we have to pull in a dep on `profile` for this :cry:

fallaciousreasoning commented on browser/ui/views/frame/split_view/brave_multi_contents_view.cc:
```diff
@@ -79,12 +90,39 @@ void BraveMultiContentsView::Layout(PassKey) {
+  // as it's attached to inactive contents view.
+  secondary_location_bar_->SetWebContents(
+      GetInactiveContentsView()->GetWebContents());
+  secondary_location_bar_->SetParentWebView(
+      contents_container_views_[inactive_index]);
```
could these lines be in an `else` block for the above `if`? Otherwise I think we set the parent web view twice (once in ctor, once here) when we construct the location bar.

fallaciousreasoning replied:
Alternatively, we could just not pass it in to the constructor? Doesn't seem like its that necessary.

fallaciousreasoning commented on browser/ui/views/split_view/split_view_location_bar.cc:
```diff
@@ -36,18 +36,17 @@
 
-  if (split_view_) {
-    view_observation_.Observe(split_view_->secondary_contents_container());
+  if (parent_web_view) {
+    view_observation_.Observe(parent_web_view);
   } else {
     CHECK_IS_TEST();
   }
```
I think we should just not pass the `parent_web_view` in here, and handle this solely from `SetParentWebView`

fallaciousreasoning reviewed (commented):
lgtm but could we add a test that the `SplitViewLocationBar` swaps back and forward on BraveMultiContentsView when the active index changes?

simonhong commented:
> lgtm but could we add a test that the `SplitViewLocationBar` swaps back and forward on BraveMultiContentsView when the active index changes?

@fallaciousreasoning Thanks for quick review! I think I addressed all your comments. PTAL again :)

goodov reviewed (approved):
`chromium_src` lgtm

simonhong commented on browser/ui/views/split_view/split_view_location_bar_browsertest.cc:
```diff
@@ -77,22 +118,30 @@ IN_PROC_BROWSER_TEST_F(SplitViewLocationBarBrowserTest,
             tab_strip_model().GetWebContentsAt(0)->GetVisibleURL().spec());
   ASSERT_EQ(u"about:blank", split_view_location_bar().url_->GetText());
 
+  // For now upstream's another page is loaded when split view is created
+  // with active tab. So, loaded newtab to use below url check.
+  if (IsSideBySideEnabled()) {
+    ASSERT_TRUE(ui_test_utils::NavigateToURLWithDisposition(
+        browser(), GURL("chrome://newtab/"), WindowOpenDisposition::CURRENT_TAB,
+        ui_test_utils::BROWSER_TEST_WAIT_FOR_LOAD_STOP));
+  }
+
```
This code is deleted by another PR(https://github.com/brave/brave-core/pull/29124)

fallaciousreasoning commented on browser/ui/views/frame/split_view/brave_multi_contents_view.h:
```diff
@@ -6,35 +6,51 @@
   static constexpr int kBorderThickness = 2;
 
-  using MultiContentsView::MultiContentsView;
+  BraveMultiContentsView(
```
out of interest, why did we need to stop with the `using` constructor? These look the same?

zeusshomaly replied:
@fallaciousreasoning do you know if you guys are working on an improved sidebar to open any website in the sidebar to the side like edge's? Brave themselves confirmed they are testing it via Twitter, see this post below, Simon is a very nice guy and is working on it actively, but I got no update about it for weeks. What's going on may I kindly ask?

Post: https://x.com/brave/status/1915448454542676027

simonhong replied:
> out of interest, why did we need to stop with the `using` constructor? These look the same?

As we have non-trivial member vars, we need ctor. Otherwise, we get below complains :) 
```
[chromium-style] Complex class/struct needs an explicit out-of-line constructor.
```

simonhong replied:
> @fallaciousreasoning do you know if you guys are working on an improved sidebar to open any website in the sidebar to the side like edge's? Brave themselves confirmed they are testing it via Twitter, see this post below, Simon is a very nice guy and is working on it actively, but I got no update about it for weeks. What's going on may I kindly ask?
> 
> Post: https://x.com/brave/status/1915448454542676027
> 
> if you do not know then I apologise, I just thought you would as you and @simonhong are working together. And I am not being impatient don't worry. I am just very excited. Edge's sidebar is so good and Brave's needs an upgrade to be like it as seen below on edge 

@zeusshomaly NP. We're going to there. but needs some more work. :)

zeusshomaly replied:
Not a problem! Thank you so much @simonhong for replaying. I was worried that's all :) I am super super excited. I cannot get it out of my head!

zeusshomaly replied:
Oh and before I forget, I suggest adding a 'sidebar apps' sync in sync so users don't have to re enter all their sidebar apps if they got a lot from one PC to another :) if you can't, no biggie

simonhong replied:
> Oh and before I forget, I suggest adding a 'sidebar apps' sync in sync so users don't have to re enter all their sidebar apps if they got a lot from one PC to another :) if you can't, no biggie

there is an issue for it https://github.com/brave/brave-browser/issues/21860

fallaciousreasoning reviewed (approved):
lgtm, with one question

brave-builds commented:
Released in [v1.80.85](https://github.com/brave/brave-browser/releases/tag/v1.80.85)
